src/mkfs.fat.c

```diff
@@ -1074,7 +1074,7 @@ static void setup_tables(void)
         }
 
         /* If is not available then generate random 32 bit disk signature */
-        if (invariant)
+        if (invariant || (getenv("SOURCE_DATE_EPOCH") && volume_id))
             disk_sig = volume_id;
         else if (!disk_sig)
             disk_sig = generate_volume_id();
```

Comment:
I'm looking at this change and I would like to know: Why check for `volume_id` is needed here? On line 514 is: `volume_id = generate_volume_id();` And `generate_volume_id()` introduced in this patch returns value from `getenv("SOURCE_DATE_EPOCH")` if this env is set.

Reply:
Hi @pali ,

> And generate_volume_id() introduced in this patch returns value from getenv("SOURCE_DATE_EPOCH") if this env is set

Yes. 
However, the _volume_id_ can be passed from "-i" option. I'm not sure if an user passed in "-i" option a null value.
We can remove the check for _volume_id_ if we make sure that "-i" cannot pass a null value.

```c
case 'i':               /* i : specify volume ID */
            errno = 0;
            conversion = strtoll(optarg, &tmp, 16);
        
            if (!*optarg || isspace(*optarg) || *tmp || conversion < 0) {
                printf("Volume ID must be a hexadecimal number\n");
```

Reply:
Well, I have not find any restriction to volume id number. So I think that **zero** is also valid volume id.

Microsoft fatgen103 specification says about volume id:
> This ID is usually generated by simply combining the current date and time into a 32-bit value.

So I think that zero volume id should be allowed.

Also FAT12 and FAT16 allows to indicate that volume id is not recorded at all and therefore allow to distinguish between _volume id is set to 0_ and _volume id is not set_.